engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniPassFactory.java

```diff
@@ -0,0 +1,29 @@
+package org.enso.compiler.pass;
+
+import org.enso.compiler.context.InlineContext;
+import org.enso.compiler.context.ModuleContext;
+
+/**
+ * A collection of factory methods for {@link MiniIRPass IR mini passes}. If a mini pass supports
+ * only inline compilation, its {@link #createForModuleCompilation(ModuleContext)} method should
+ * return null.
+ */
+public interface MiniPassFactory<T extends MiniIRPass> {
+  /**
+   * Creates an instance of mini pass that is capable of transforming IR elements in the context of
+   * a module.
+   *
+   * @param moduleContext A mini pass can optionally save reference to this module context.
+   * @return May return null if module compilation is not supported.
+   */
+  T createForModuleCompilation(ModuleContext moduleContext);
```

Comment:
You don't need a type variable. Just return `MiniPIRPass`.

Reply:
Removed type parameter from `MiniIrPass` in 28edccb6c6fb1b73c46a65e964984d396ecf0b63. Parametrizing mini passes based on their metadata types would make chaining difficult.